### PR TITLE
Reflect to proxy object, to observe changes made by setters

### PR DESCRIPTION
**Pull request by tomalec**
For example, to be able to observe the change made to `obj.foo` by setting `obj.bar='new'`
```js
let obj = {
  'foo': 'old'
};
Object.defineProperty(obj, 'bar', {
  set: function (newValue) {
    this.foo = newValue;
  },
  get: function () {
    return this.foo
  },
  enumerable: true
});
```

Before this PR, `this` in setter was referring to original object not the proxy.

This change decreases performance significantly. According to `npm run bench` executed 3 times on my local it's:
```
jsonpatcherproxy generate operation -20,97%
jsonpatch generate operation -6,31%
jsonpatcherproxy mutation - huge object -5,12%
jsonpatch mutation - huge object -5,21%
jsonpatcherproxy generate operation - huge object -9,18%
jsonpatch generate operation - huge object -1,73%
PROXIFY big object -1,05%
DIRTY-OBSERVE big object -5,67%
```

### CODE REVIEW

**Comment by alshakero** [comment]
[React Easy State](https://github.com/solkimicreb/react-easy-state) does not make this sacrifice. 

See "Avoid using the this keyword in the methods of your state stores."

https://github.com/solkimicreb/react-easy-state

**Comment by warpech** [comment]
It was discussed on a HO between me and @tomalec that we should introduce 2 modes of operation in JSONPatcherProxy:

- fast that does not reflect the proxy object to get and set operations. In fact, it does not need to use the Reflect API at all, it can use a quicker method: `tree[key] = newValue`.
- slow that reflects the proxy object to get and set operations. It must use the Reflect API

I can volunteer to do this change, but only if PR #42 is accepted first. Otherwise I would go nuts with the tests :)

File: test/spec/proxySpec.js [hunk 1]
```diff
@@ -173,6 +173,43 @@ describe('proxy', function() {
...
+        'bar': 'old'
+      };
+
+      jsonpatch.applyPatch(obj2, patches);
+
+      /* iOS and Android */
+      observedObj = JSONPatcherProxy.deepClone(observedObj);
+
+      expect(obj2).toReallyEqual(observedObj);
+    });
```
[the remark this answers is not in the document]
Comment (tomalec) [reply test/spec/proxySpec.js]: Well they check if the patch application leads to the correct result. However they are quire redundant with `fast-json-patch` tests.

**Comment by warpech** [comment]
This PR was rebased on current `master` and is now available as a new PR #44

**Pull request closed without merging**